content/news/2022-04-12-bevy-0.7/index.md

```diff
@@ -0,0 +1,383 @@
+fn system(mut set: QuerySet<(QueryState<(&mut A, &B)>, QueryState<(&mut A, &C)>)>) {
+    for (a, b) in set.q0().iter_mut() {
+    }
+}
+```
+
+**Bevy 0.7** removes `QuerySet` in favor of `ParamSet`, which generalizes the QuerySet pattern for _any_ system parameter:
+
+```rust
+fn system(mut set: ParamSet<(Query<(&mut A, &B)>, Query<(&mut A, &C)>)>) {
+    for (a, b) in set.p0().iter_mut() {
+    }
+}
+```
+
+But ParamSets aren't just limited to Queries! Consider this example, where the `EventWriter<Jump>` parameter (which internally accesses the `Events<Jump>` resource) conflicts with the raw access to that resource. Previously, expressing this wouldn't be possible. But with ParamSets, it is!
+
+```rust
+fn system(mut set: ParamSet<(EventWriter<Jump>, ResMut<Events<Jump>>)>) {
+    for jump_event in set.p1().drain() {
+    }
+}
+```
+
+We still recommend avoiding ParamSets where possible for clarity's sake. But every so often they are a necessary and useful tool!
+
+## Deref / DerefMut Derives
+
+<div class="release-feature-authors">authors: @MrGVSV</div>
+
+Rust encourages the use of the [newtype pattern](https://doc.rust-lang.org/rust-by-example/generics/new_types.html) when expanding a type with new functionality or meaning. This is also a useful tool in Bevy:
+
+```rust
+#[derive(Component)]
+struct Items(Vec<Item>);
+
+fn give_sword(mut query: Query<&mut Items>) { 
+    for mut items in query.iter_mut() {
+        items.0.push(Item::new("Flaming Poisoning Raging Sword of Doom"));
+    }
+}
+```
+
+This works just fine, but that `0` at the end of `items.0` sticks out like a sore thumb. Many of us in the Bevy Org think `.0` has no place in public apis. But the newtype pattern is still useful! Ideally, Rust would provide a way to express that `Items` is a new type, while transparently provided access to the `Vec<Item>` stored within. There are designs being discussed by the Rust team, but we don't want to wait for nice things!
+
+Fortunately, the Deref / DerefMut traits in std provide the behavior we want. Users can already manually implement these traits, but for such a common pattern, we decided that providing our own trait derives was worth it. In **Bevy 0.7**, you can now derive Deref and DerefMut, enabling much nicer public apis:
+
+```rust
+#[derive(Component, Deref, DerefMut)]
+struct Items(Vec<Item>);
+
+fn give_sword(mut query: Query<&mut Items>) { 
+    for mut items in query.iter_mut() {
+        // No more .0!
+        items.push(Item::new("Flaming Poisoning Raging Sword of Doom"));
+    }
+}
+```
+
+Astute `std` doc readers might notice that the Rust team [recommends only using Deref/DerefMut for smart pointers, to avoid confusion](https://doc.rust-lang.org/std/ops/trait.Deref.html). Components like `Items` _are not_ smart pointers. We personally don't care. This pattern works, it is already widely used in the Rust ecosystem, and Good UX Comes First.
```

Comment:
```suggestion
Astute `std` doc readers might notice that the Rust team [recommends only using `Deref`/`DerefMut` for smart pointers, to avoid confusion](https://doc.rust-lang.org/std/ops/trait.Deref.html). Components like `Items` _are not_ smart pointers. We choose to ignore this advice, as this pattern works, is already widely used in the Rust ecosystem, and Good UX Comes First.
```

Mostly, I really don't like:
> We personally don't care.

For a number of reasons.

Reply:
Yea, this is needlessly inflammatory.

Also, if you consider me as part of the "bevy org", you are not speaking for me, as i personally do care and i am against (ab)using deref in this way.

Reply:
Solid points all around. I'm happy to own this decision personally if that sits better with you all (I just default to "we" when discussing features and design choices generally, even when I've fully designed and implemented something). And toning down the language is a very positive improvement here that better reflects the tone we should be projecting.

Reply:
(I just committed @DJMcNab's suggestion, but ill leave this open until we make a call on the "we" thing)

Reply:
I think "as a project, we've chosen to..." is good language here. Not everyone agrees with the call, but in the context of Bevy as a project / organization, we've made a decision.